apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/TimedBufferingConsumer.cs

```diff
@@ -61,26 +64,77 @@ public static TimedBufferingConsumer<T> Create(IConsumer<T> consumer, TimeSpan w
             => new TimedBufferingConsumer<T>(consumer, waitTime, new SimpleThreadingTimer());
 
         /// <inheritdoc />
-        public override void Dispose()
+        public void Receive(IEnumerable<T> items)
         {
-            _timer.Dispose();
-            Flush();
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            var enumerator = items.GetEnumerator();
+            if (enumerator.MoveNext())
+            {
+                Receive(enumerator);
+            }
         }
 
         /// <inheritdoc />
-        protected override bool ReceiveWithSemaphoreHeld(IEnumerable<T> items)
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken)
         {
-            GaxPreconditions.CheckNotNull(items, nameof(items));
-            _items.AddRange(items);
-            return false;
+            // TODO: Should we do something with cancellationToken, like check for cancellation
+            // on every iteration?
+            Receive(items);
+            return Task.CompletedTask;
+        }
+
+        private void Receive(IEnumerator<T> remainingItems)
+        {
+            try
+            {
+                do
+                {
+                    // We very explicitly use _items here instead of a local
+                    // reference. On Flush, _items will be assigned a new
+                    // collection and we want to use the new collection as soon
+                    // as possible.
+                    // Note that it's still possible that
+                    // 1. We read _items
+                    // 2. We loose execution and a Flush starts.
+                    // 3. Flush swaps _items.
+                    // 4. Flush marks the old items (the one we are referencing) as CompleteAdding.
+                    // 5. We gain execution and execute Add on a completed collection, which fails.
+                    // That's why we catch InvalidOperationException and simply reattempt just
+                    // where we left off.
+                    // Also note that if we regain execution in between 3 and 4, that's fine.
+                    // Our call to Add will happen before the call to CompleteAdding, so it will
+                    // succeed before the collection is marked as complete.
+                    _items.Add(remainingItems.Current);
+                }
+                while (remainingItems.MoveNext());
+            }
+            catch (InvalidOperationException)
+            {
+                Receive(remainingItems);
```

Comment:
I'd like us to have some guard on the recursion here. If we end up recursing (say) 10 levels deep, I think we should fail with an exception or similar - rather than going potentially forever until a StackOverflowException brings down the whole process.

Reply:
OK, I'll reset it on each success and let's see what happens.